log/log.go

```diff
@@ -0,0 +1,220 @@
+package log
+
+import (
+	"github.com/getsentry/raven-go"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+var (
+	logger = zap.NewNop().Sugar()
+)
+
+// Level is the verbose representation of log level
+type Level string
+
+// Enums for Level
+const (
+	NopLevel   Level = "nop"
+	DebugLevel Level = "debug"
+	InfoLevel  Level = "info"
+	WarnLevel  Level = "warn"
+	ErrorLevel Level = "error"
+	PanicLevel Level = "panic"
+	FatalLevel Level = "fatal"
+
+	// This will have the same effect as nop but slower
+	ZapNopLevel zapcore.Level = zapcore.FatalLevel + 1
+)
+
+// InitLogger provides a quick way to start or replace a logger
+// Pass in debug as log level enables development mode (which makes DPanicLevel logs panic).
+// Optionally pass in a cfg will provide a logger with custom setting
+func InitLogger(logLevel Level, cfg ...zap.Config) {
+	if logLevel == NopLevel {
+		logger = zap.NewNop().Sugar()
+		return
+	}
+
+	var config zap.Config
+
+	if len(cfg) > 0 {
+		config = cfg[0]
+	} else {
+		config = zap.NewProductionConfig()
+		{
+			lvl := StringToAtomicLevel(logLevel)
+			config.Encoding = "console"
+			config.EncoderConfig.EncodeCaller = ShortCallerEncoder
+			config.EncoderConfig.EncodeTime = TimeEncoder
+			config.EncoderConfig.EncodeLevel = CapitalLevelEncoder
+			config.Level = zap.NewAtomicLevelAt(lvl)
+
+			if lvl == zap.DebugLevel {
+				config.Development = true
+			}
+		}
+	}
+
+	// TODO: error handling
+	l, _ := config.Build(zap.AddCallerSkip(2))
+	logger = l.Sugar()
+}
+
+// StringToAtomicLevel converts in to a zap acceptable atomic level struct
+func StringToAtomicLevel(loglevel Level) zapcore.Level {
+	lvl := ZapNopLevel
+
+	switch loglevel {
+	case DebugLevel:
+		lvl = zapcore.DebugLevel
+	case InfoLevel:
+		lvl = zapcore.InfoLevel
+	case WarnLevel:
+		lvl = zapcore.WarnLevel
+	case ErrorLevel:
+		lvl = zapcore.ErrorLevel
+	case PanicLevel:
+		lvl = zapcore.PanicLevel
+	case FatalLevel:
+		lvl = zapcore.FatalLevel
+	}
+
+	return lvl
+}
```

Comment:
nit: I think this is the same as:
```go
switch loglevel {
default:
	return ZapNopLevel
case DebugLevel:
	return zapcore.DebugLevel
case InfoLevel:
	return zapcore.InfoLevel
case WarnLevel:
	return zapcore.WarnLevel
case ErrorLevel:
	return zapcore.ErrorLevel
case PanicLevel:
	return zapcore.PanicLevel
case FatalLevel:
	return zapcore.FatalLevel
}
```
which is slightly clearer.

Reply:
maybe put `default` at the bottom? In most cases Nop is least used case.

Reply:
refactored and moved the default case to the bottom.

Reply:
🔕 I don't think `default` being at the top or bottom will actually have a performance impact. When it's at the top we still need to compare against all other cases before falling back to `default`. You can check the assembly code it generated to verify that :)

I personally slightly prefer to have `default` at the top in go code as a way to make it more obvious that this `switch...case` block do have a `default` branch, but I don't feel strongly about that.